Error with `numpy >= 2.4.0`

Ran into this today - no time to look further into it, but the new numpy broke vigra.

```
__________________ ERROR collecting test/test_arraytypes.py ___________________
test_arraytypes.py:44: in <module>
    import vigra.arraytypes as arraytypes
..\vigra\__init__.py:113: in <module>
    from . import arraytypes
..\vigra\arraytypes.py:340: in <module>
    class VigraArray(numpy.ndarray):
..\vigra\arraytypes.py:1438: in VigraArray
    @_preserve_doc
     ^^^^^^^^^^^^^
..\vigra\arraytypes.py:49: in _preserve_doc
    npy_doc = eval('numpy.ndarray.%s.__doc__' % f.__name__)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
<string>:1: in <module>
    ???
E   AttributeError: type object 'numpy.ndarray' has no attribute 'ptp'
------------------------------- Captured stderr -------------------------------
```

xref: https://github.com/conda-forge/vigra-feedstock/issues/155

--- a/docsrc/credits_changelog.dxx
+++ b/docsrc/credits_changelog.dxx
@@ -87,6 +87,14 @@
 Many thanks to all!
 <p>
 
+<b> Changes from Version 1.12.2 to 1.12.3 (2025/03/13)</b>
+
+<ul>
+
+      <li> Vigra Numpy: fix compatibility with numpy 2.4. The methods ptp of the Vigra arrays are conditionally removed if numpy 2.4 or greater is detected.
+
+</ul>
+
 <b> Changes from Version 1.12.1 to 1.12.2 (2025/03/13)</b>
 
 <ul>

--- a/vigranumpy/lib/arraytypes.py
+++ b/vigranumpy/lib/arraytypes.py
@@ -37,6 +37,7 @@ from functools import reduce
 import sys
 import copy
 import numpy
+
 from . import ufunc
 import vigra.vigranumpycore as vigranumpycore
 
@@ -44,6 +45,7 @@ from vigra.vigranumpycore import AxisType, AxisInfo, AxisTags
 
 from collections.abc import Iterable
 
+numpyHasPtp = numpy.__version__.startswith(("1.", "2.0.", "2.1.", "2.2.", "2.3."))
 
 def _preserve_doc(f):
     npy_doc = eval('numpy.ndarray.%s.__doc__' % f.__name__)
@@ -1435,22 +1437,23 @@ class VigraArray(numpy.ndarray):
         '''
         return _numpyarray_overloaded_function(numpy.ndarray.prod, self, axis, dtype, out)
 
-    @_preserve_doc
-    def ptp(self, axis=None, out=None):
-        '''
-        The 'axis' parameter can be an int (axis position) or string (axis key).
-        '''
-        if type(axis) == str:
-            axis = self.axistags.index(axis)
-        if axis is None:
-            return numpy.ptp(self.transposeToOrder('C').view(numpy.ndarray), out=out)
-        else:
-            res = numpy.ptp(self.view(numpy.ndarray), axis=axis, out=out)
-            if out is None:
-                res = res.view(VigraArray)
-                res.axistags = self._copy_axistags()
-                del res.axistags[axis]
-            return res
+    if numpyHasPtp:
+        @_preserve_doc
+        def ptp(self, axis=None, out=None):
+            '''
+            The 'axis' parameter can be an int (axis position) or string (axis key).
+            '''
+            if type(axis) == str:
+                axis = self.axistags.index(axis)
+            if axis is None:
+                return numpy.ptp(self.transposeToOrder('C').view(numpy.ndarray), out=out)
+            else:
+                res = numpy.ptp(self.view(numpy.ndarray), axis=axis, out=out)
+                if out is None:
+                    res = res.view(VigraArray)
+                    res.axistags = self._copy_axistags()
+                    del res.axistags[axis]
+                return res
 
     @_preserve_doc
     def ravel(self, order='C'):

--- a/vigranumpy/lib/axistags.py
+++ b/vigranumpy/lib/axistags.py
@@ -1,4 +1,5 @@
 import numpy as np
+numpyHasPtp = np.__version__.startswith(("1.", "2.0.", "2.1.", "2.2.", "2.3."))
 
 
 class TaggedArray(np.ndarray):
@@ -88,11 +89,12 @@ class TaggedArray(np.ndarray):
             del res.axistags[axis]
         return res
 
-    def ptp(self, axis=None, out=None):
-        res = np.ndarray.ptp(self, axis, out)
-        if axis is not None:
-            del res.axistags[axis]
-        return res
+    if numpyHasPtp:
+        def ptp(self, axis=None, out=None):
+            res = np.ndarray.ptp(self, axis, out)
+            if axis is not None:
+                del res.axistags[axis]
+            return res
 
     def ravel(self, order='C'):
         res = np.ndarray.ravel(self, order)

--- a/vigranumpy/lib/tagged_array.py
+++ b/vigranumpy/lib/tagged_array.py
@@ -36,6 +36,8 @@
 import copy, sys
 import numpy
 
+numpyHasPtp = numpy.__version__.startswith(("1.", "2.0.", "2.1.", "2.2.", "2.3."))
+
 
 def preserve_doc(f):
     f.__doc__ = eval('numpy.ndarray.%s.__doc__' % f.__name__)
@@ -263,13 +265,14 @@ of 'numpy.ndarray'.
             del res.axistags[axis]
         return res
 
-    @preserve_doc
-    def ptp(self, axis=None, out=None):
-        res = numpy.ndarray.ptp(self, axis, out)
-        if axis is not None:
-            res.axistags = res.copy_axistags()
-            del res.axistags[axis]
-        return res
+    if numpyHasPtp:
+        @preserve_doc
+        def ptp(self, axis=None, out=None):
+            res = numpy.ndarray.ptp(self, axis, out)
+            if axis is not None:
+                res.axistags = res.copy_axistags()
+                del res.axistags[axis]
+            return res
 
     @preserve_doc
     def ravel(self, order='C'):

--- a/vigranumpy/test/test_arraytypes.py
+++ b/vigranumpy/test/test_arraytypes.py
@@ -51,7 +51,7 @@ import platform
 IS_PYPY = platform.python_implementation() == 'PyPy'
 
 
-numpyHasComplexNegateBug = numpy.version.version.startswith('1.0')
+numpyHasPtp = numpy.__version__.startswith(("1.", "2.0.", "2.1.", "2.2.", "2.3."))
 
 try:
     vt.testAny()
@@ -1471,8 +1471,9 @@ def testMethods():
     assert_equal(ones.prod(), 1.0)
     assert (ones.prod(axis='y') == [1]*ones.shape[0]).all()
 
-    assert_equal(a.ptp(), a.size-1)
-    assert (a.ptp(axis='x') == [a.shape[0]-1]*a.shape[1]).all()
+    if numpyHasPtp:
+        assert_equal(a.ptp(), a.size-1)
+        assert (a.ptp(axis='x') == [a.shape[0]-1]*a.shape[1]).all()
 
     r = arraytypes.ScalarImage((2,2))
     r.ravel()[...] = range(4)
@@ -1555,8 +1556,7 @@ def testUfuncs():
         b = -a
         assert_equal(t, b.dtype)
         assert_equal(a.axistags, b.axistags)
-        if not numpyHasComplexNegateBug or t is not clongdouble:
-            assert (b == -t(2)).all()
+        assert (b == -t(2)).all()
         b = a + a
         assert_equal(t, b.dtype)
         assert_equal(a.axistags, b.axistags)

